### ISSUE TITLE
Animations in iOS 9 broken?

### ISSUE THREAD

**Issue by bmueller** [PureLayout/PureLayout#82]
Has anyone had issues with UIView animations of PureLayout constraints with the iOS 9 beta? 

After upgrading to the beta, I noticed that every time I tried to animate a position change for certain constraints, the position change happened instantly with no animation whatsoever. This didn't happen for all constraints that I tried to animate - some animated just like on iOS 8, others did not. 

Eventually I realized that the constraints that weren't animating correctly were set up with PureLayout, while the ones that were working were hooked up in Interface Builder.

Here's the animation code that works fine for some objects, and worked fine for everything in iOS 8:

self.balloonConstraint.constant = -70;
[UIView animateWithDuration:5.0 animations:^{
       [self.balloon layoutIfNeeded];
}];

I could just switch to using Interface Builder for these objects, but hopefully this is just a bug that will be fixed before GM - or something got changed that I'm unaware of, and I have to change a little code.

**Comment by smileyborg** [comment]
Hey @bmueller,

I am not aware of any issue like this (with PureLayout or auto layout in general on iOS 9). As you know, PureLayout is just a simple convenient wrapper around the regular auto layout SDK API. It doesn't do anything fancy or create special constraints under the hood. So I doubt that this is a PureLayout-specific issue -- if it is an issue, it likely affects constraints generated or activated in code in general.

Regardless, I'd love to help dig in to what's happening here. The small code snippet you posted looks fine to me. Are you able to share a small sample project that reproduces the issue? Or can you modify one of the PureLayout demos in the example app to reproduce it? Also, does this happen on the device or simulator or both (I assume you're using the latest iOS 9 beta 5)?

**Comment by bmueller** [comment]
Yeah, I figured this likely wasn't related specifically to PureLayout, but this seemed like a good first place to ask just in case there was some weird iOS 9 bug I was unaware of.

Just did a little more digging after seeing that the sample project animations work fine. Comparing it to my code, the sample code in ALiOSDemo7.m calls layoutIfNeeded on self.view instead of the individual views, which is what I was doing. When I change it to [self.redView layoutIfNeeded] and run it in the simulator, the animations break exactly as I was describing.

I ran this test in the simulator on iOS 9 beta 5. (But I also saw this happening in my project on-device.)

**Comment by smileyborg** [comment]
OK, I definitely see the change in behavior modifying the iOS example app Demo 7:
* On iOS 8, this code:

  ```
  animations:^{
      [self.view layoutIfNeeded];
  }
  ```

  produces the same result as this code:

  ```
  animations:^{
      [self.redView layoutIfNeeded];
      [self.blueView layoutIfNeeded];
  }
  ```

* On iOS 9, this code produces the same results as both of the above:

  ```
  animations:^{
      [self.view layoutIfNeeded];
  }
  ```

  but this code produces **no animation whatsoever** and the views jump between positions:

  ```
  animations:^{
      [self.redView layoutIfNeeded];
      [self.blueView layoutIfNeeded];
  }
  ```

The significant difference here is calling `layoutIfNeeded` on the view whose position is affected by a change in constraints, versus calling `layoutIfNeeded` on a **superview** of the view.

This is definitely a change between how iOS 8 and 9 work. (It does make some sense to me why calling `layoutIfNeeded` on the superview works, because the constraints that affect the position of subviews are added to the superview.) I can't say for sure whether it's *supposed* to work like it does in iOS 8, though -- maybe you're only supposed to call `layoutIfNeeded` on the superview of views you wish to animate, and it was just by luck that it worked before.

I would definitely [file a radar](http://bugreport.apple.com) with Apple on this (I will file one too).

**Comment by bmueller** [comment]
Glad I'm not going crazy! :)

The weird thing is that calling [subview layoutIfNeeded] works just fine if the constraints were set up in Interface Builder - this only happens with constraints set up in code. That's why I originally thought this might be a bug with PureLayout and came here first, but I doubt it has anything to do with PureLayout.

I'll file a radar now, though at least I was able to fix the issue in my code by animating the superview instead. Thanks for pointing me towards the fix!

**Comment by smileyborg** [comment]
That is very interesting that it works if the constraints/views come from IB. I don't have any good theory on why that would be the case, but clearly constraints are handled a bit differently when coming from IB.

I would definitely point this out in the radar.

**Comment by smileyborg** [comment]
I created a simple sample project to attach to the radar to reproduce this issue: http://cl.ly/2f0q1l0z3F1B

**Comment by smileyborg** [comment]
Filed this as rdar://22345582

**Comment by zsll** [comment]
I have the same issue with my swift 2 code on IOS 9 as well. My code is as below:
```
    func animateViewMoving (up:Bool, moveValue :CGFloat){
        let movementDuration:NSTimeInterval = 0.3
        let movement:CGFloat = ( up ? -moveValue : moveValue)
        UIView.beginAnimations( "animateView", context: nil)
        UIView.setAnimationBeginsFromCurrentState(true)
        UIView.setAnimationDuration(movementDuration )
        self.view.frame = CGRectOffset(self.view.frame, 0,  movement)
        UIView.commitAnimations()
    }
```
The behavior is like it works for a moment but bounces back right away.

**Comment by liuxuan30** [comment]
I would like to know if there any dashboard can see what's changed for auto-layout implementation on iOS 9. @smileyborg are you aware of any?

**Comment by BrysovVadik** [comment]
I had a problem when I tried to get programmatically the width and height constraints, all others constraints worked correctly. I solved the problem by adding [[layoutConstraint class] isEqual:[NSLayoutConstraint class]] in method:
- (NSLayoutConstraint *)sz_heightConstraint {
    for (NSLayoutConstraint *layoutConstraint in self.constraints) {
        if([[layoutConstraint class] isEqual:[NSLayoutConstraint class]] && layoutConstraint.firstAttribute == NSLayoutAttributeHeight) {
            return layoutConstraint;
        }
    }
    return nil;
}

The problem is that ios9 added other subclass in layoutConstraint.

**Comment by johnlui** [comment]
As @smileyborg said, make it's super view to layoutIfNeeded can fix this problem caused by iOS 9.

**Comment by smileyborg** [comment]
Here's what I heard back from Apple:

> It was always a mistake to send -layoutIfNeeded to the view that was moving instead of its superview.

So, this may or may not be "fixed" in iOS 9. Sounds like the fact that it worked in previous versions of iOS was just luck.

Anyways, closing this issue, as it's really nothing to do with PureLayout and we have no control over if or how it gets fixed.

**Comment by antonbremer** [comment]
"It was always a mistake to send -layoutIfNeeded to the view that was moving instead of its superview."
Just guessing here: Probably because when you change constraints of a view, it might cause changes in other views as well, therefore you need to call the superview layoutIfNeeded?

**Comment by jgurnani** [comment]
I had also experienced the similar issue but noticed some thing different. I found that suppose we are animating two views (lets say A and B) and the view whose constraints are with the view (some other view ) that is not animating like super view or else is animating fine. But the view whose constraints are set with respect to view that is animating (Say A or B) is not animated properly. I don't know you found the same or not.
